prov/net/src/xnet_progress.c

```diff
@@ -94,6 +119,9 @@ static ssize_t xnet_send_msg(struct xnet_ep *ep)
 
 	assert(xnet_progress_locked(xnet_ep2_progress(ep)));
 	assert(ep->cur_tx.entry);
+	if (xnet_io_uring && !xnet_get_uring_credit(&xnet_ep2_progress(ep)->tx_uring))
+		return -FI_EAGAIN;
```

Comment:
Thanks for the review. 
We can provision the ring to be larger than the max ep's we want to support, but I have some concerns with this strategy: 
- the net provider sets `ep_cnt` to 8192, but it doesn't seem that the code actually restricts the number of endpoints.
- we will need to provision extra uring entries to post cancellation requests of pending SQEs when an endpoint is flushed (not implemented yet).

I guess the question is: do we want to flow-control io_uring or do we want to always provision the io_uring for the worst case scenario?
Personally, I like the idea where the number of uring entries is independent of the maximum number of endpoints. I would expect the system to reach its peek performance after a certain number of SQEs inflight, which obviously depends on the number of communicating EPs, but not on the max EPs allocated. However, adding a flow-control might be a problem for fairness as the last EPs in the TX/RX lists may not always be served because the first EPs in the list may have consumed all the entries already. Maybe your suggestion is the best approach after all. 

BTW, the flag [IORING_FEAT_NODROP](https://manpages.debian.org/testing/liburing-dev/io_uring_setup.2.en.html#IORING_FEAT_NODROP) could also be used to bring some kind of control flow. When this flags is set, `io_uring_submit()` would return -EBUSY if the ring is full and cannot accept new SQEs.

Reply:
I'd prefer the explicit check, because I can envision needing to add it at some point anyway to debug some issue.  It seems that moving the credit check closer to io_uring_submit() would align with the behavior of NODROP.  I don't know what performance penalty we'd hit if NODROP is enabled and we hit that situation frequently.  The description makes it sound like NODROP might be slower than just not overrunning the cq.

Reply:
@shefty There is one problem if we move the credit check closer to io_uring_submit(): it is impossible to select how many ready SQEs will be submitted: it's either all or none. If the provider has prepared too many SQEs, it has to wait for at least as many free credits as the number of SQEs prepared before it can call `io_uring_submit()`. 
I decided to move the credit check under the sockapi. Please let me know your opinion.